### PR TITLE
feat: add preliminary local checks for forc-publish

### LINKED ISSUE #7252: forc-publish should move some sanity checks to client side
Currently `forc-publish` does not check version, depenedency states (whether they are by path or version or git) and relies the backend to do these checks, we should make an effort on the plugin side to early error out and have these checks on the plugin side as well.

### PR DESCRIPTION
**Pull request by kayagokalp**
## Description
closes #7252.

This pull request introduces enhancements to the `forc-publish` plugin, focusing on validating project configurations and improving error handling. Key changes include the addition of a `validate` module for preemptive checks, updates to error definitions, and modifications to the tarball creation logic to incorporate validation. New test cases and sample data directories have also been added to ensure robust validation functionality.

### Validation Enhancements:
* Added a new `validate` module in `forc-publish/src/validate.rs` to check for missing `Forc.toml`, missing version fields, and unsupported dependency types.
* Updated the `Error` enum in `forc-publish/src/error.rs` to include new error variants for validation failures like `InvalidForcToml`, `MissingVersionField`, and `DependencyMissingVersion`.

### Tarball Creation Updates:
* Refactored `create_tarball_from_current_dir` in `forc-publish/src/tarball.rs` to use the new `validate_dir` function for preemptive validation. Removed redundant checks for `Forc.toml` existence.
* Adjusted test cases in `forc-publish/src/tarball.rs` to use sample directories with predefined configurations for validation testing. [[1]](diffhunk://#diff-e3f20cadab3a1cf99fc967e0adc12e908f435a185e1529df9b152388927adf45L79-R85) [[2]](diffhunk://#diff-e3f20cadab3a1cf99fc967e0adc12e908f435a185e1529df9b152388927adf45L140-R135)

### Test Data Additions:
* Added sample directories under `tests/data` for various validation scenarios, including projects with missing versions, dependencies without versions, and valid configurations. [[1]](diffhunk://#diff-b171d6da23a1fe36735cd962cfcf506831115c3714654236b31dd999c2853823R1-R11) [[2]](diffhunk://#diff-016348eaea4a95d041537b29d370619ab0d5ca99cc84f9b30d8ad561aa757a1eR1-R9) [[3]](diffhunk://#diff-0495f782769450a8c200fdcc1f29805f4811a26f4849720fa34d5b37ba45e471R1-R8)
* Included `.gitignore` files in test data directories to exclude `out` and `target` folders.

### CODE REVIEW

**Review by JoshuaBatty — DISMISSED** [review]

File: forc-plugins/forc-publish/src/error.rs [hunk 1]
```diff
@@ -29,6 +29,18 @@ pub enum Error {
     #[error("Forc.toml not found in the current directory")]
     ForcTomlNotFound,
 
+    #[error("Invalid forc.toml: {0}")]
+    InvalidForcToml(#[from] anyhow::Error),
+
+    #[error("Project is missing a version field, add one under [project]")]
```
Comment (zees-dev) [inline forc-plugins/forc-publish/src/error.rs]: nit:

```suggestion
    #[error("missing version field in [project] section of Forc.toml")]
```

File: forc-plugins/forc-publish/src/tarball.rs [hunk 2]
```diff
@@ -156,6 +151,8 @@ mod test {
 
         // Run the function
         let result = create_tarball_from_current_dir(&temp_output_dir);
+        result.unwrap();
```
Comment (zees-dev) [inline forc-plugins/forc-publish/src/tarball.rs]: redundant call to the same function?

**Review by zees-dev — DISMISSED** [review]
LGTM 👍 

Potential improvements:
I think additional test(s) can be added to test the following edge cases:
- Empty `Forc.toml` files
- Malformed TOML syntax
- Dependencies with empty version strings

Could be separate PR.

File: forc-plugins/forc-publish/src/error.rs (hunk 1 above)
Comment (kayagokalp) [reply forc-plugins/forc-publish/src/error.rs]: Thanks!, done in [1bdcba4](https://github.com/FuelLabs/sway/pull/7253/commits/1bdcba4084831f8004423022aef19d63a8cfd1bf)

File: forc-plugins/forc-publish/src/tarball.rs (hunk 2 above)
Comment (kayagokalp) [reply forc-plugins/forc-publish/src/tarball.rs]: Fixed in [1bdcba4](https://github.com/FuelLabs/sway/pull/7253/commits/1bdcba4084831f8004423022aef19d63a8cfd1bf), thanks for catching this!

**Comment by NICO41NIKO** [comment]
nice
